Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -173,22 +179,28 @@ void GrpcStream::FinishAndNotify(const Status& status) {
 }
 
 void GrpcStream::Shutdown() {
+  LOG_DEBUG(
+      "GrpcStream('%s'): shutting down; completion(s): %s, is finished: %s",
+      this, completions_.size(), is_grpc_call_finished_);
+
   MaybeUnregister();
-  if (completions_.empty()) {
-    // Nothing to cancel -- either the call was already finished, or it has
-    // never been started.
-    return;
+
+  if (!completions_.empty() && !is_grpc_call_finished_) {
+    // Important: since the stream always has a pending read operation,
```

Comment:
This comment is hard to parse because:
  * it conflates pre- and post-conditions
  * doesn't clearly indicate normal and error operation
  * uses ambiguous phrasing (e.g. "cancellation" could seemingly apply to either the context or the call)

Consider phrasing something like:

> Important: during normal operation, the stream always has a pending read operation so `Shutdown` would hang indefinitely if we didn't cancel the context_. However, if the stream has already failed, avoid canceling the context to avoid overwriting the status captured during the `OnOperationFailed`.

Reply:
Thanks, I like this wording.